### DISCUSSION TITLE
Tables in PGML

### DISCUSSION THREAD

**Discussion by Alex-Jordan** [openwebwork/pg#909] [Ideas]
Seeking advice, most likely from @dpvc. With `niceTables.pl`, one of a few ways to construct a table is like:

```
DataTable(
[                            # array ref of the rows
    [                        # start a row
        ['A', options],      # a cell
        ['B', options],      # a cell
    ],
    [                        # start a row
        ['C', options],      # a cell
        ['D', options],      # a cell
    ],
],
table options
)
```

This way offers enough structure to put options on the cells. Note that `DataTable` has a `LayoutTable` counterpart.

I would like to make PGML markup that mimics this and is ultimteley translated to the above. I would like it to be such that for cells, an author still writes content using PGML markup. For example a cell might have `[_]{$answer}` in it. Here is the markup for a DataTable that I have gotten started with, for within PGML.

```
[#
    [-
        [.A.]{options}
        [.B.]{options}
    -]
    [-
        [.C.]{options}
        [.D.]{options}
    -]
#]{table options}
```

and for a LayoutTable, a starred variant: `[# ... #]*{table options}`. All the newlines and indentation above is just for presentation, and is not supposed to make a difference with output. For example, an author could make the less readable:
```
[# [- [.[`x`].]{headerrow => 1}
      [.[`x^2`].] -]
   [- [.[`2`].]
      [.[_]{4}.] -] #]{align => '|c|c|', midrules => 1, center => 0}
```

I want to ask for thoughts on this markup. Some thoughts/observations I had:
* In the `[# ... #]` I really only want to parse rows. But it seems like with machinery that is there now, I need to `parseAll`.
* Similarly within a `[- ... -]` I really only want to parse cells.
* I've hit some issues with using `[- ... -]` where the closing `-]` gets parse as a horizontal rule. Was dash an intrinsically bad choice in any obvious way?

Should I be using block delimiters with brackets in the first place? Given the `niceTables.pl` structure, what markup would you come up with for this?

**Comment by dpvc** [comment]
You are right that doing all the things you want would take some changes to the infrastructure, but it doesn't take too much change to make at least a preliminary version of this possible.  I've made a version of PGML.pl called [PGMLtables.pl](https://gist.github.com/dpvc/1e52c01aba0ab9452bb4c006cec35cb7) that includes the needed changes.

The main structural different in the PGMLtables.pl file is that the block definitions now allow you to specify a `PGML::Block` subclass to use when creating a block definition.  This makes it possible to override the `pushItem()` method so that you can check that only rows can be added to a table, and only cells to a row.  The remainder of the changes define the implementation for tables.

One caveat is that the contents of the cells can't be block-level commands (like paragraph breaks and numbered lists).  It may be possible to allow that, but I have run out of time to work on this, and haven't looked into it to see how hard that would be (frankly, I have forgotten how it all works, and it is complicated).  For now, you can't use those within a cell, and if you do, the results are completely messed up.  That should be addressed in the future.

Because indentation is meaningful in PGML (indentation by 4 spaces produces indentation in the output), allowing the indentation in your tables is a bit awkward.  I have added a flag that ignores indentation commands, and that does allow for indentation, but it does mean you can't add indentation in the cells.  But since you can't currently do block-level commands in the cells anyway, indenting would not be available anyway.  It may be possible in the future to allow it (by removing the indentation of the `[.` to determine the actual indentation) when block-level layout is available in the cells.

Anyway, the link above should give you a `PGMLtables.pl` file that does the trick.  You can use `diff` to see the differences from the original `PGML.pl` (I used the one currently in develop as the starting point).

**Comment by dpvc** [comment]
PS, I don't seem to have any problems with `-]`, but I don't know how you defined things in your copy of `PGML.pl`.  In any case, this seems to work for me.

**Comment by dpvc** [comment]
I've updated the `PGMLtables.pl` file linked above to include support for block-level element inside table cells.  It wasn't all that hard to do it, but I haven't done much testing, so there may be hidden gotchas.  Let me know if you find any of those.

**Comment by Alex-Jordan** [comment]
Thanks Davide, this will be great. I have been bogged down with local things but I want to try this out and put it to use soon. Thank you for your time on this.
